### ISSUE TITLE
USB pass-through (storage, audio, webcam, scanner and other devices - like a second mouse)

### ISSUE THREAD

**Issue by Torinde** [joncampbell123/dosbox-x#3210]
USB pass-through (storage, audio, webcam, scanner and other devices - like a second mouse) from host into DOSbox-X. Mostly useful for Win9x, but also can help with for dual mouse games. [Here is shown an example set of USB drivers for DOS - USBUHCI, USBMOUSE, CuteMouse](https://retrocomputing.stackexchange.com/questions/17700/use-a-usb-mouse-for-dos-games)

_Originally posted by @Torinde in https://github.com/joncampbell123/dosbox-x/issues/2412#issuecomment-1008162065_

**Comment by joncampbell123** [comment]
That would be cool, but to get there, DOSBox-X will first need USB host controller emulation (as a PCI device of course).

http://hackipedia.org/browse.cgi/Computer/Platform/PC%2c%20IBM%20compatible/Busses/USB%2c%20Universal%20Serial%20Bus/Host%20controller/UHCI%20Universal%20Host%20Controller%20Interface%20Design%20Guide%20%281996%2d03%29%20v1%2e1%2epdf

http://hackipedia.org/browse.cgi/Computer/Platform/PC%2c%20IBM%20compatible/Busses/USB%2c%20Universal%20Serial%20Bus/Host%20controller/OHCI%20Open%20Host%20Controller%20Interface%20Specification%20for%20USB%20%281999%2d09%2d14%29%20v1%2e0a%2epdf

**Comment by Torinde** [comment]
VirtualBox does that - user can connect/disconnect any USB device from the host and use it in the guest - and afterwards, when disconnected from guest the device re-connects to the host. I don't know if some code can be reused from there?

Also related - scanner/webcam emulation from picture/video file (not sure of the use case though).

**Comment by Torinde** [comment]
[PCem doesn't have it yet](https://pcem-emulator.co.uk/phpBB3/viewtopic.php?p=14125) as well. https://github.com/sarah-walker-pcem/pcem/issues/136

Related:
- https://github.com/86Box/86Box/discussions/2907

**Comment by Torinde** [comment]
Virtualbox - [The EHCI and XHCI USB controller devices are now part of the open source base package](https://www.virtualbox.org/wiki/Changelog-7.0#v00)

[Virtualbox 7.2.2](https://www.virtualbox.org/wiki/Changelog-7.2#v2):
> USB Webcam: The virtual USB webcam is now part of the open source base package

QEMU and Bochs also support USB emulation.

**Comment by Torinde** [comment]
@crazii, how will your https://github.com/crazii/USBDDOS fit with such a feature? Have you seen any potential implementations for "USB host controller emulation"?

**Comment by crazii** [comment]
I don't think so. It is merely a dos driver that won't support win9x guest and no host emulation at all.

**Comment by Torinde** [comment]
There are external USB-to-ISA adapters [1](http://arstech.com/install/ecom-prodshow/usb2isar.html) [2](https://www.arstech.com/install/cms-display/ste_usl.html) - they even sell [dosbox-x.conf and win9x.img](https://arstech.com/install/index.php?app=ecom&ns=prodshow&ref=img-wME-w98) files! [I wonder](https://github.com/joncampbell123/dosbox-x/issues/3269#issuecomment-1037106568) how they get access to the USB-attached ISA slots in DOSbox-X - my assumptions are either:
- by modifying DOSbox-X source code to provide ISA pass-trough (**can we see that source?**) and then making their own Win7/10 USB ISA driver (which [USB Device class](https://en.wikipedia.org/wiki/USB#Device_classes)?)
- by modifying DOSbox-X source code to provide USB pass-trough (**can we see that source?**) and then making their own DOS USB ISA driver (like https://github.com/crazii/USBDDOS/issues/1)

The second option will be relevant for this Issue, although I think the first is more likely (because they mention a Win7/10 ISA SDK/software layer).

**Comment by rderooy** [comment]
Where does it say that those config files are meant for use with those ISA to USB adapters?

My understanding is that they are selling a pre-installed Win98SE/WinME OS image + dosbox-x config files.

**Comment by Torinde** [comment]
So, first it's stated that along with the USB2ISA you need to consider power, enclosure, cable, software. 

For the software it's written than "installation software" is sufficient, but "if you have access to the ISA software source code" you can buy an extra SDK. Elsewhere they write that they "offer support for DOS under Widows operating systems".

Then, there are the same 4 options to choose: power enclosure, cable, software. Here the software is "OS images for dosbox-x" consisting of DOSbox-X.conf and Win9x.IMG. True that they don't even state DOSbox-X itself is included.

But still, the question remains - how do you access those ISA-over-USB ports in DOSbox?
